File: src/Q219/Solution.java
package Q219;

import java.util.HashMap;

public class Solution {
    public boolean containsNearbyDuplicate(int[] nums, int k) {
        HashMap<Integer, Integer> map = new HashMap<>(nums.length);
        for (int i = 0; i < nums.length; i++) {
            Integer old = map.put(nums[i], i);
            if (old != null && i - old <= k) {
                return true;
            }
        }
        return false;
    }

    public static void main(String[] args) {
        Solution q219 = new Solution();
        int[] nums = new int[]{1,2,3,1,2,3};
        int k = 2;
        System.out.println(q219.containsNearbyDuplicate(nums,k));
    }
}

File: src/Q219/Solution.py
from typing import List

"""
Given an integer array nums and an integer k, return true if there are two distinct indices i and j in the array 
such that nums[i] == nums[j] and abs(i - j) <= k.
"""

class Solution:
    def containsNearbyDuplicate(self, nums: List[int], k: int) -> bool:
        curr = set()
        for right in range(len(nums)):
            if right-k-1 >= 0:
                curr.remove(nums[right-k-1])
            if nums[right] in curr:
                return True
            curr.add(nums[right])
        return False


if __name__ == '__main__':
    s = Solution()
    print(s.containsNearbyDuplicate([1,2,3,1], 3))
    print(s.containsNearbyDuplicate([1,0,1,1], 1))
    print(s.containsNearbyDuplicate([1,2,3,1,2,3], 2))